src/masternodes/poolpairs.cpp

```diff
@@ -0,0 +1,190 @@
+    }
+    else if (poolPairByTokens && (poolPairByTokens->second.idTokenA != pool.idTokenA || poolPairByTokens->second.idTokenB == pool.idTokenB))
+    {
+        throw std::runtime_error("Error, idTokenA or idTokenB is incorrect.");
+    }
+
+    return Res::Err("Error: Couldn't create/update pool pair.");
+}
+
+Res CPoolPairView::UpdatePoolPair(DCT_ID const & poolId, bool & status, CAmount const & commission, CScript const & ownerAddress)
+{
+    auto poolPair = GetPoolPair(poolId);
+    if (!poolPair) {
+        return Res::Err("Pool with poolId %s does not exist", poolId.ToString());
+    }
+
+    CPoolPair & pool = poolPair.get();
+
+    if (pool.status != status) {
+        pool.status = status;
+    }
+    if (commission >= 0) { // default/not set is -1
+        if (commission > COIN) {
+            return Res::Err("commission > 100%%");
+        }
+        pool.commission = commission;
+    }
+    if (!ownerAddress.empty()) {
+        pool.ownerAddress = ownerAddress;
+    }
+
+    auto res = SetPoolPair(poolId, pool);
+    if (!res.ok) {
+        return Res::Err("Update poolpair: %s" , res.msg);
+    }
+    return Res::Ok();
+}
+
+boost::optional<CPoolPair> CPoolPairView::GetPoolPair(const DCT_ID &poolId) const
+{
+    DCT_ID poolID = poolId;
+    return ReadBy<ByID, CPoolPair>(WrapVarInt(poolID.v));
+}
+
+boost::optional<std::pair<DCT_ID, CPoolPair> > CPoolPairView::GetPoolPair(const DCT_ID &tokenA, const DCT_ID &tokenB) const
+{
+    DCT_ID poolId;
+    auto varint = WrapVarInt(poolId.v);
+    ByPairKey key {tokenA, tokenB};
+    if(ReadBy<ByPair, ByPairKey>(key, varint)) {
+        auto poolPair = ReadBy<ByID, CPoolPair>(varint);
+        if(poolPair)
+            return { std::make_pair(poolId, std::move(*poolPair)) };
+    }
+    return {};
+}
+
+Res CPoolPair::Swap(CTokenAmount in, PoolPrice const & maxPrice, std::function<Res (const CTokenAmount &tokenAmount)> onTransfer) {
+    if (in.nTokenId != idTokenA && in.nTokenId != idTokenB) {
+        throw std::runtime_error("Error, input token ID (" + in.nTokenId.ToString() + ") doesn't match pool tokens (" + idTokenA.ToString() + "," + idTokenB.ToString() + ")");
+    }
+    if (in.nValue <= 0)
+        return Res::Err("Input amount should be positive!");
+
+    if (!status)
+        return Res::Err("Pool trading is turned off!");
+
+    bool const forward = in.nTokenId == idTokenA;
+
+    // it is important that reserves are at least SLOPE_SWAP_RATE (1000) to be able to slide, otherwise it can lead to underflow
+    if (reserveA < SLOPE_SWAP_RATE || reserveB < SLOPE_SWAP_RATE)
+        return Res::Err("Lack of liquidity.");
+
+    auto aReserveA = arith_uint256(reserveA);
+    auto aReserveB = arith_uint256(reserveB);
+
+    arith_uint256 maxPrice256 = arith_uint256(maxPrice.integer) * PRECISION + maxPrice.fraction;
+    arith_uint256 priceAB = (aReserveA * PRECISION / aReserveB);
+    arith_uint256 priceBA = (aReserveB * PRECISION / aReserveA);
```

Comment:
Before divide need check aReserveA is 0?

Reply:
No. See `if (reserveA < SLOPE_SWAP_RATE || reserveB < SLOPE_SWAP_RATE)` upthere